hypercall.h

```diff
@@ -10,11 +10,12 @@ int strnlen_n(char* s, int max_len){
 
 // #if defined(__x86_64__)
 #if defined(__x86_64__)
-static inline int hc(char *s) {
+static inline int hc(char **s,int len) {
     uint64_t eax = MAGIC_VAL;
     uint64_t ret = MAGIC_VAL;
-    uint64_t action = (uint32_t)strnlen_n(s, 0x1000);
-
+    for(int i = 0; i< len; i++){
+        strnlen_n(s[i],0x1000);
+    }
```

Comment:
Here's a different design that could work and potentially run faster. I didn't actually test this, just sketching it out. Example hc2 function is just for mips32.

```cpp
static inline uint32_t hc2(uint32_t num, uint32_t a1) {
  // Run a hypercall with two arguments and consume a return value that emulator should place
  // into register R5
  uint32_t result;
  register uint32_t a1reg asm("$5") = a1;
  asm volatile(
    "movz $0, %[num], $5\n"
    "move %[result], $5"
    : [result] "=r" (result)
    : [num] "r" (num), "r"(a1reg)
    );
    return result;
}

// Function to reliably inform our pyplugin that an event has occurred with the specified buffer of length len
void hc_str(int event, char* buffer, int len) {
 uint32_t pending_page_in = (uint32_t)-1);
 while (pending_page_in != 0) {
    if (pending_page_in != (uint32_t)-1) {
      strlen_n(pending_page_in, 0x100); // Page in the panda-requested virtual memory
    }
    hc2(event, buffer)
  }
}
```

Now, from the pypanda side of things, we can control if and when we page in guest memory by returning values we can't read

```py3
events = {0x123456: 'nvram_read', ...}
@panda.cb_guest_hypercall
def hypercall(cpu):
    event = panda.arch.get_arg(cpu, 0, convention="syscall")
    if event not in events:
        return
        guest_pointer= panda.arch.get_arg(cpu, 1, convention="syscall")
    try:
        buffer = panda.read_str(cpu, guest_ptr, 0x1000)
    except ValueError:
        # Find paged out memory
        for i in range(0, 0x1000, 0x10)
            try:
                c = panda.virtual_memory_read(cpu, guest_ptr+i, 0x10)
              except ValueError:
                 panda.arch.set_reg(cpu, 0, convention='syscall', guest_ptr+i) # Failed to read, make guest page it in
                 return
    # Now we have the full buffer paged in

```